README.rst
*******
CurlyTx
*******
Enigma2 (Dreambox) plugin that lets you view the contents of remote
*plain text* files, e.g. HTTP URLs.

Multiple URLs can be configured and navigated.

.. contents::
   :depth: 2


========
Features
========
- Load any remote plain text files, e.g. via HTTP
- Unlimited number of remote URLs/pages
- Reload pages
- Non-blocking URL loading
- Show HTTP headers
- Configurable page titles
- Configurable text size
- Configurable default page
- Import complete page list from Atom feed
- Visible in the main menu or the extension menu (configurable)
- Configurable menu title
- Help screen for main window and settings window


=====
Usage
=====

First run
=========
After installing CurlyTx and restarting Enigma2, open the main menu.
The first entry will be "CurlyTx" - activate it.

You will see the main window with the message
"Go and add a page in the settings".
Do just that end press the red button to access the settings window.

Now we'll add the first URL:

- Press the yellow button ("New"); the "page edit" window will show up
- Enter the page URL, e.g. http://ip.cweiske.de
- If you wish, enter a page title, e.g. "My IP"
- Set the text size if you want. 20 is a good default value.
- Press the green button ("OK"), and you are back on the settings window.
- The page you have just created is in the configuration list now.


If you made a mistake and want to change it, select the page with the
up/down buttons and press "OK" - the page edit window will open.

Press the green button and the settings will be saved.
You're back on the main window now and the URL you just configured will be loaded.

Note that with version 1.1, CurlyTx supports opening local files.
Prepend them with ``file://``, e.g. ``file:///tmp/test.txt``.


Adding many pages
=================
You can use the settings window to add new pages, but this gets tedious if you
want to add many pages.

It's better to use the Atom feed import in this case.
All you need is a text editor and a web server you can serve the feed page with.

Here is an example feed::

  <?xml version="1.0" encoding="utf-8"?>
  <feed xmlns="http://www.w3.org/2005/Atom">
   <title>URL list for CurlyTx</title>
   <author>
    <name>Christian Weiske</name>
    <email>cweiske@cweiske.de</email>
   </author>
   <link rel="self" href="http://home.cweiske.de/pagefeed.atom"/>
   <entry>
    <id>ip</id>
    <title>My IP</title>
    <link rel="alternate" type="text/html" href="http://ip.cweiske.de/" />
   </entry>
   <entry>
    <id>temp</id>
    <title>House temperatures</title>
    <link rel="alternate" type="text/html" href="http://home/temperatures.txt" />
   </entry>
  </feed>

Start CurlyTx, go to the settings and write the feed URL in the
"Page feed URL" field.
Then press "OK" and the feed's pages will be loaded into the settings window.


Local configuration file
========================
CurlyTx looks for a local page feed file on the harddisk on every run,
at ``/etc/enigma2/curlytx-pagefeed.xml``.
If this file exists, it gets loaded unconditionally and overwrites
the current page configuration.


Disable settings
================
The page feed file may contain a tag that disables the settings button.
By using it, you can prevent people from modifying the CurlyTx settings.

First, register the namespace::

    <feed xmlns="http://www.w3.org/2005/Atom" xmlns:c="http://ns.cweiske.de/curlytx">

Then, add the setting after the author or self link::

    <c:enableSettings>0</c:enableSettings>

Here is the example feed with disabled settings::

  <?xml version="1.0" encoding="utf-8"?>
  <feed xmlns="http://www.w3.org/2005/Atom" xmlns:c="http://ns.cweiske.de/curlytx">
   <title>URL list for CurlyTx</title>
   <author>
    <name>Christian Weiske</name>
    <email>cweiske@cweiske.de</email>
   </author>
   <link rel="self" href="http://home.cweiske.de/pagefeed.atom"/>
   <c:enableSettings>0</c:enableSettings>
   <entry>
    <id>ip</id>
    <title>My IP</title>
    <link rel="alternate" type="text/html" href="http://ip.cweiske.de/" />
   </entry>
   <entry>
    <id>temp</id>
    <title>House temperatures</title>
    <link rel="alternate" type="text/html" href="http://home/temperatures.txt" />
   </entry>
  </feed>


=================
Modifying CurlyTx
=================


Translation
===========
Beginning a new translation
---------------------------
Replace ``$lang_code`` with your two-letter language code::

    $ cd po
    $ cp messages.po $lang_code.po
    ... edit $lang_code.po now
    $ ./compile.sh


Editing an existing translation
-------------------------------
Simply run ::

    $ cd po
    $ ./update.sh

This will update the translation template ``messages.pot`` from the source code
and will merge the changes into the single translation files.


Testing a translation
---------------------
Link your compiled translation file into ::

    src/locale/$lang_code/LC_MESSAGES/CurlyTx.mo

Enigma2 will pick it up automatically.


Building
========
First upgrade the version number in ``CONTROL/control``.

Then simply run ::

    ./build.sh

Directory ``releases/`` will contain the freshly baked ``.ipk`` file that can
then be transferred to your dreambox, e.g. via ``scp``::

    $ scp releases/enigma2-plugin-extensions-curlytx_2.3_mipsel.ipk dreambox:
    $ ssh dreambox
    $ ipkg install enigma2-plugin-extensions-curlytx_2.3_mipsel.ipk

You will need the ``ipkg-build`` script from
 http://reichholf.net/files/dreambox/tools/ipkg-build

Also see http://dream.reichholf.net/wiki/Howto:IPK_Pakete_erstellen


Open issues / ideas
===================
- move mode to re-order pages
- how to show clock in lcd?


=====
About
=====
Homepage
========
https://open-dreambox.org/trac/dreambox/wiki/CurlyTx

Author
======
Christian Weiske, `cweiske@cweiske.de`__

.. __: mailto:cweiske@cweiske.de

License
=======
The plugin is subject to the GPLv3_ or later.

Additional exception:
  This plugin may be distributed and executed on hardware which is licensed by
  Dream Multimedia GmbH.

.. _GPLv3: http://www.gnu.org/licenses/agpl.html


ChangeLog
2014-01-30  Christian Weiske  <cweiske@cweiske.de>

	* Support older dreamboxes (e.g. 7025+) that don't have a
	  recent elementtree library.
	* Version 1.1.1

2013-09-02  Christian Weiske  <cweiske@cweiske.de>

	* Add support for loading local files via "file://" URLs
	* Add support for local configuration file
	* Add option to disable settings button, load option from XML page feed file
	* Fix crash on help screen
	* Show CurlyTx in plugin browser
	* Version 1.1.0

2011-12-06  Christian Weiske  <cweiske@cweiske.de>

	* Fix crash that occured when the box is slow
	* Fix crash showing headers when no headers available
	* Fix bug: Saving URLs was broken
	* Show error message when feed loading fails
	* Add usage instructions to README
	* Add docblocks to the code
	* Error messages can be translated now
	* Version 1.0

2011-11-24  Christian Weiske  <cweiske@cweiske.de>

	* Show headers when pressing info button
	* Version 0.4

2011-11-23  Christian Weiske  <cweiske@cweiske.de>

	* Add page list import via Atom feed

2011-11-21  Christian Weiske  <cweiske@cweiske.de>

	* Hide color buttons when no/only one page configured
	* Hide default page config option when no page defined
	* Show current page when multiple ones are defined: "2/6"
	* Configurable text font size per page

2011-11-18  Christian Weiske  <cweiske@cweiske.de>

	* Add help screens
	* Fix fixed size menu title setting
	* Add messages.pot to ipkg distribution
	* Translations work now
	* Version 0.3

2011-11-17  Christian Weiske  <cweiske@cweiske.de>

	* Version 0.2


src/config.py
# -*- coding: utf-8 -*-
# CurlyTx configuration
# Copyright (C) 2011 Christian Weiske <cweiske@cweiske.de>
# License: GPLv3 or later

from Components.config import config, ConfigEnableDisable, ConfigYesNo, ConfigSelection, ConfigNumber, ConfigText, ConfigSubsection, ConfigSubList, ConfigInteger

def createPage():
    """ Create and return a configuration page object """
    s = ConfigSubsection()
    s.uri   = ConfigText(default="http://", fixed_size=False)
    s.title = ConfigText(
        default = "Page #" + str(len(config.plugins.CurlyTx.pages) + 1),
        fixed_size = False
        )
    s.fontSize = ConfigInteger(20, (1, 100))
    return s

def loadDefaultPageOptions():
    defaults = []
    for i in range(0, len(config.plugins.CurlyTx.pages)):
        defaults.append((str(i), config.plugins.CurlyTx.pages[i].title.value))
    if hasattr(config.plugins.CurlyTx, "defaultPage"):
        config.plugins.CurlyTx.defaultPage.setChoices(defaults, "0")
    else:
        config.plugins.CurlyTx.defaultPage = ConfigSelection(defaults, "0")

def feedPagesToConfig(pages):
    """ save pages from atom feed into config. """
    if len(pages) == 0:
        return

    del config.plugins.CurlyTx.pages[:]

    for pageData in pages:
        page = createPage()
        config.plugins.CurlyTx.pages.append(page)
        page.title.setValue(pageData["title"])
        page.uri.setValue(pageData["url"])

def feedSettingsToConfig(settings):
    changed = False
    if 'enableSettings' in settings and config.plugins.CurlyTx.enableSettings.getValue() != settings['enableSettings']:
        config.plugins.CurlyTx.enableSettings.setValue(int(settings['enableSettings']))
        changed = True

    if changed:
        config.plugins.CurlyTx.save()

def savePageConfig():
    for i in range(0, len(config.plugins.CurlyTx.pages)):
        config.plugins.CurlyTx.pages[i].save()

    config.plugins.CurlyTx.pages.save()


#configuration setup
config.plugins.CurlyTx = ConfigSubsection()
config.plugins.CurlyTx.menuMain = ConfigYesNo(default = True)
config.plugins.CurlyTx.menuExtensions = ConfigYesNo(default = False)
config.plugins.CurlyTx.enableSettings = ConfigEnableDisable(default = True)
config.plugins.CurlyTx.menuTitle = ConfigText(default = "CurlyTx", fixed_size = False)
config.plugins.CurlyTx.feedUrl = ConfigText(default = "", fixed_size = False)
config.plugins.CurlyTx.pages = ConfigSubList()
for id,value in config.plugins.CurlyTx.pages.stored_values.iteritems():
    config.plugins.CurlyTx.pages.append(createPage())
loadDefaultPageOptions()


src/plugin.py
# -*- coding: utf-8 -*-
# CurlyTx plugin registration and startup
# Copyright (C) 2011 Christian Weiske <cweiske@cweiske.de>
# License: GPLv3 or later

from Plugins.Plugin import PluginDescriptor

import CurlyTx
from . import config
from Components.config import config

 
def main(session, **kwargs):
    """ Opens the main window """
    try:
        session.open(CurlyTx.CurlyTx)
    except:
        import traceback
        traceback.print_exc()

def menuHook(menuid):
    """ Called whenever a menu is created """
    if menuid == "mainmenu" and config.plugins.CurlyTx.menuMain.value:
        return [(config.plugins.CurlyTx.menuTitle.value, main, "curlytx", 41)]
    return [ ]

 
def Plugins(**kwargs):
    """ Register CurlyTx in the extension list and main menu """
    list = [
        PluginDescriptor(name = "CurlyTx",
                         description = "View remote text files",
                         where = [PluginDescriptor.WHERE_PLUGINMENU],
                         fnc = main),
        PluginDescriptor(name = config.plugins.CurlyTx.menuTitle.value,
                         description = _("View remote text files"),
                         where=PluginDescriptor.WHERE_MENU,
                         fnc = menuHook),
        ]
    if config.plugins.CurlyTx.menuExtensions.value:
        list.append(
            PluginDescriptor(name = config.plugins.CurlyTx.menuTitle.value,
                             description = _("View remote text files"),
                             where = [PluginDescriptor.WHERE_EXTENSIONSMENU],
                             fnc = main)
        )
    return list


src/AtomFeed.py
# -*- coding: utf-8 -*-
# CurlyTx Atom feed parser
# Copyright (C) 2011 Christian Weiske <cweiske@cweiske.de>
# License: GPLv3 or later

from twisted.web.client import getPage
from xml.etree.cElementTree import fromstring

import os

class AtomFeed:
    """ Simple XML parser that extracts pages from a atom feed """
    ns = "{http://www.w3.org/2005/Atom}"
    nsc = "{http://ns.cweiske.de/curlytx}"
    errorCallback = None

    def __init__(self, url, callback, errorCallback):
        """ Fetches the URL

        Parsed pages are sent back to callback by parse()
        """
        self.errorCallback = errorCallback
        if (url.startswith('file://')):
            file = url[7:]
            if not os.path.exists(file):
                errorCallback('Settings atom feed file does not exist: ' + file)
                return

            with open(file, 'r') as f:
                self.parse(f.read(), callback)
        else:
            getPage(url).addCallback(self.parse, callback).addErrback(self.onError)

    def parse(self, data, callback):
        """ Parse atom feed data into pages list and run callback """
        try:
            xml = fromstring(data)
        except Exception:
            return self.errorCallback("Invalid XML")

        pages = []
        for entry in xml.findall("{0}entry".format(self.ns)):
            titleE = entry.find("{0}title".format(self.ns))
            url   = self.bestLink(entry.findall("{0}link".format(self.ns)))
            if titleE != None and titleE.text != "" and url != None:
                pages.append({"title": titleE.text, "url": url})

        settings = dict()
        for entry in list(xml):
            if (entry.tag.startswith(self.nsc)):
                settings[entry.tag[len(self.nsc):]] = entry.text

        callback(pages, settings)

    def bestLink(self, list):
        """ Fetch the best matching link from an atom feed entry """
        foundLevel = -1
        foundHref = None
        for link in list:
            if link.get("rel") != "alternate" and link.get("rel") != "":
                continue
            level = self.level(link)
            if foundLevel > level:
                continue
            foundLevel = level
            foundHref = link.get("href")
        return foundHref

    def level(self, link):
        """ Determines the level of a link

        "text/plain" type links are best, links without type are second.
        All others have the lowest level 1.
        """
        type = link.get("type")
        if type == "text/plain":
            return 3
        elif type == "":
            return 2
        return 1

    def onError(self, error):
        """ Pass the error message only """
        self.errorCallback(error.getErrorMessage())


src/__init__.py
# -*- coding: utf-8 -*-
# CurlyTx translation initialization
# Copyright (C) 2011 Christian Weiske <cweiske@cweiske.de>

from Components.Language import language
from Tools.Directories import resolveFilename, SCOPE_PLUGINS, SCOPE_LANGUAGE
from os import environ as os_environ
import gettext

def localeInit():
    """ Prepare settings for gettext usage """
    lang = language.getLanguage()[:2] # getLanguage returns e.g. "fi_FI" for "language_country"
    os_environ["LANGUAGE"] = lang # Enigma doesn't set this (or LC_ALL, LC_MESSAGES, LANG). gettext needs it!
    gettext.bindtextdomain("CurlyTx", resolveFilename(SCOPE_PLUGINS, "Extensions/CurlyTx/locale"))

def _(txt):
    """ Custom gettext translation function that uses the CurlyTx domain """
    t = gettext.dgettext("CurlyTx", txt)
    if t == txt:
        #print "[CurlyTx] fallback to default translation for", txt
        t = gettext.gettext(txt)
    return t

localeInit()
language.addCallback(localeInit)


src/CurlyTxSettings.py
# -*- coding: utf-8 -*-
# CurlyTx configuration window
# Copyright (C) 2011 Christian Weiske <cweiske@cweiske.de>
# License: GPLv3 or later

from . import _

from Screens.Screen import Screen
from Screens.HelpMenu import HelpableScreen
from Components.ActionMap import ActionMap, NumberActionMap
from Components.Sources.StaticText import StaticText
from Screens.MessageBox import MessageBox

from . import config
from config import createPage, loadDefaultPageOptions, feedPagesToConfig, feedSettingsToConfig, savePageConfig
from Components.config import config, getConfigListEntry, ConfigSelection
from Components.ConfigList import ConfigList, ConfigListScreen

class CurlyTxSettings(ConfigListScreen, HelpableScreen, Screen):
    skin = """
	<screen name="Setup" position="center,center" size="560,430" title="Settings">
	  <ePixmap pixmap="skin_default/buttons/red.png"    position="0,0"   size="140,40" transparent="1" alphatest="on" />
	  <ePixmap pixmap="skin_default/buttons/green.png"  position="140,0" size="140,40" transparent="1" alphatest="on" />
	  <ePixmap pixmap="skin_default/buttons/yellow.png" position="280,0" size="140,40" transparent="1" alphatest="on" />
	  <ePixmap pixmap="skin_default/buttons/blue.png"   position="420,0" size="140,40" transparent="1" alphatest="on" />
	  <widget source="key_red"    render="Label" position="0,0"   zPosition="1" size="140,40" font="Regular;20" halign="center" valign="center" backgroundColor="#9f1313" transparent="1" />
	  <widget source="key_green"  render="Label" position="140,0" zPosition="1" size="140,40" font="Regular;20" halign="center" valign="center" backgroundColor="#1f771f" transparent="1" />
	  <widget source="key_yellow" render="Label" position="280,0" zPosition="1" size="140,40" font="Regular;20" valign="center" halign="center" backgroundColor="#a08500" transparent="1" />
	  <widget source="key_blue"   render="Label" position="420,0" zPosition="1" size="140,40" font="Regular;20" valign="center" halign="center" backgroundColor="#18188b" transparent="1" />
	  <widget name="config" position="5,50" size="550,325" scrollbarMode="showOnDemand" />
	</screen>"""

    def __init__(self, session):
        self.skin = CurlyTxSettings.skin
        Screen.__init__(self, session)
        HelpableScreen.__init__(self)
        #self.skinName = [ "CurlyTxSettings", "Setup" ]
        self.setup_title = _("Settings")

        self["actions"] = ActionMap(["SetupActions","ColorActions"],
            {
                "cancel": self.keyCancel,
                "save": self.keySave,
                "ok": self.editPage,
                "yellow": self.newPage,
                "blue": self.deletePage
            }, -2)

        self["key_red"]    = StaticText(_("Cancel"))
        self["key_green"]  = StaticText(_("OK"))
        self["key_yellow"] = StaticText(_("New"))
        self["key_blue"]   = StaticText(_("Delete"))

        ConfigListScreen.__init__(self, self.getConfigList(), session = self.session)

        self.loadHelp()

    def getConfigList(self):
        #reload titles
        loadDefaultPageOptions()
        cfg = config.plugins.CurlyTx

        list = [
            getConfigListEntry(_("Page:") + " " + x.title.value, x.uri)
                for x in cfg.pages
            ]
        if len(cfg.pages):
            list.append(getConfigListEntry(_("Default page"), cfg.defaultPage))
        list.append(getConfigListEntry(_("Show in main menu"), cfg.menuMain))
        list.append(getConfigListEntry(_("Show in extensions menu"), cfg.menuExtensions))
        list.append(getConfigListEntry(_("Menu title"), cfg.menuTitle))
        list.append(getConfigListEntry(_("Page feed URL"), cfg.feedUrl))
        return list

    def loadHelp(self):
        self.helpList.append((
                self["actions"], "SetupActions",
                [("cancel", _("Dismiss all setting changes"))]))
        self.helpList.append((
                self["actions"], "SetupActions",
                [("save", _("Save settings and close screen"))]))
        self.helpList.append((
                self["actions"], "SetupActions",
                [("ok", _("Edit selected page"))]))
        self.helpList.append((
                self["actions"], "SetupActions",
                [("ok", _("Load pages from feed"))]))
        self.helpList.append((
                self["actions"], "ColorActions",
                [("yellow", _("Add new page"))]))
        self.helpList.append((
                self["actions"], "ColorActions",
                [("blue", _("Delete selected page"))]))

    def keyLeft(self):
        ConfigListScreen.keyLeft(self)

    def keyRight(self):
        ConfigListScreen.keyRight(self)

    def deletePage(self):
        pageCount = len(config.plugins.CurlyTx.pages)
        if pageCount == 0 or self["config"].getCurrentIndex() >= pageCount:
            return

        from Screens.MessageBox import MessageBox
        self.session.openWithCallback(
            self.deletePageConfirm,
            MessageBox,
            _("Really delete this page?\nIt cannot be recovered!")
            )

    def deletePageConfirm(self, result):
        if not result:
            return

        id = self["config"].getCurrentIndex()
        del config.plugins.CurlyTx.pages[id]

        config.plugins.CurlyTx.pages.save()

        self["config"].setList(self.getConfigList())

    def newPage(self):
        from CurlyTxSettings import CurlyTxSettings
        self.session.openWithCallback(self.pageEdited, CurlyTxPageEdit, createPage(), True)

    def editPage(self):
        id = self["config"].getCurrentIndex()
        if id < len(config.plugins.CurlyTx.pages):
            self.session.openWithCallback(
                self.pageEdited, CurlyTxPageEdit,
                config.plugins.CurlyTx.pages[id], False
                )
        elif config.plugins.CurlyTx.feedUrl.value:
            from AtomFeed import AtomFeed
            AtomFeed(
                config.plugins.CurlyTx.feedUrl.value,
                self.feedPagesReceived, self.feedPagesFail
                )
        else:
            self.session.open(
                MessageBox, _("No page feed URL defined"), MessageBox.TYPE_ERROR
                )

    def pageEdited(self, page, new):
        if not page:
            return

        if new:
            config.plugins.CurlyTx.pages.append(page)

        self["config"].setList(self.getConfigList())

    def feedPagesReceived(self, pages, settings):
        feedPagesToConfig(pages)
        feedSettingsToConfig(settings)
        self["config"].setList(self.getConfigList())

    def feedPagesFail(self, errorMessage):
        """ Downloading the page url feed failed somehow """
        self.session.open(
            MessageBox,
            _("Error loading page feed:") + "\n\n" + str(errorMessage),
            MessageBox.TYPE_ERROR
            )

    def keySave(self):
        savePageConfig()
        ConfigListScreen.keySave(self)

    def cancelConfirm(self, result):
        """Overwriting ConfigListScreen.cancelConfirm to call cancelAll method"""
        if not result:
            return

        self.cancelAll()
        self.close()

    def cancelAll(self):
        for x in self["config"].list:
            x[1].cancel()

        #restore old page configuration
        cfg = config.plugins.CurlyTx
        del cfg.pages[:]
        for i in cfg.pages.stored_values:
            cfg.pages.append(createPage())



class CurlyTxPageEdit(Screen, ConfigListScreen):
    def __init__(self, session, page, new = False):
        Screen.__init__(self, session)
        self.skinName = [ "CurlyTxPageEdit", "Setup" ]

        self["key_red"]   = StaticText(_("Cancel"))
        self["key_green"] = StaticText(_("OK"))

        self["setupActions"] = ActionMap(["SetupActions"],
            {
                "save": self.save,
                "cancel": self.keyCancel
	    }, -1)

        self.page = page
        self.new = new
        list = [
            getConfigListEntry(_("Page URL"), page.uri),
            getConfigListEntry(_("Title"), page.title),
            getConfigListEntry(_("Font size"), page.fontSize),
            ]
        ConfigListScreen.__init__(self, list, session = self.session)

    def save(self):
        self.close(self.page, self.new)

    def keyCancel(self):
        self.close(None, self.new)


src/CurlyTx.py
# -*- coding: utf-8 -*-
# CurlyTx main window
# Copyright (C) 2011 Christian Weiske <cweiske@cweiske.de>
# License: GPLv3 or later

from . import _

from Screens.Screen import Screen
from Screens.HelpMenu import HelpableScreen
from Components.Label import Label
from Components.ScrollLabel import ScrollLabel
from Components.ActionMap import ActionMap
from Components.Sources.StaticText import StaticText
from Tools import Directories
from twisted.web import client
from twisted.web.client import _makeGetterFactory, HTTPClientFactory
from enigma import gFont

from . import config
from config import createPage, loadDefaultPageOptions, feedPagesToConfig, feedSettingsToConfig, savePageConfig
from Components.config import config

import os

class CurlyTx(Screen, HelpableScreen):
    skin = """
        <screen name="CurlyTx" position="center,center" size="560,430" title="CurlyTx" >
	  <ePixmap position="0,0" size="140,40" pixmap="skin_default/buttons/red.png" transparent="1" alphatest="on" />
	  <ePixmap position="140,0" size="140,40" pixmap="skin_default/buttons/green.png" transparent="1" alphatest="on" />
	  <ePixmap position="280,0" size="140,40" pixmap="skin_default/buttons/yellow.png" transparent="1" alphatest="on" />
	  <ePixmap position="420,0" size="140,40" pixmap="skin_default/buttons/blue.png" transparent="1" alphatest="on" />
	  <widget source="key_red" render="Label" position="0,0" zPosition="1" size="140,40" valign="center" halign="center" font="Regular;21" transparent="1" foregroundColor="white" />
	  <widget source="key_green" render="Label" position="140,0" zPosition="1" size="140,40" valign="center" halign="center" font="Regular;21" transparent="1" foregroundColor="white" />
	  <widget source="key_yellow" render="Label" position="280,0" zPosition="1" size="140,40" valign="center" halign="center" font="Regular;21" transparent="1" foregroundColor="white" />
	  <widget source="key_blue" render="Label" position="420,0" zPosition="1" size="140,40" valign="center" halign="center" font="Regular;21" transparent="1" foregroundColor="white" />
	  <widget name="text" position="5,45" size="550,380" font="Console;20" />
        </screen>"""

    currentUrl = None
    currentPage = None
    currentFontSize = 20
    httpGetterFactory = None
    showingHeaders = False
    staticPageFeedFile = 'curlytx-pagefeed.xml'

    def __init__(self, session, args = None):
        #self.skin = CurlyTx.skin
        Screen.__init__(self, session)
        HelpableScreen.__init__(self)
        #self.skinName = [ "CurlyTx", "Setup" ]

        self["text"] = ScrollLabel("foo")

        self["key_red"]    = StaticText(_("Settings"))
        self["key_green"]  = StaticText(_("Reload"))
        self["key_yellow"] = StaticText(_("Prev"))
        self["key_blue"]   = StaticText(_("Next"))


        self["actions"] = ActionMap(
            ["WizardActions", "ColorActions", "InputActions", "InfobarEPGActions"], {
                "ok":   self.close,
                "back": self.close,
                "up":   self.pageUp,
                "down": self.pageDown,

                "red":    self.showSettings,
                "green":  self.reload,
                "yellow": self.prevPage,
                "blue":   self.nextPage,

                "showEventInfo": self.showHeader
            }, -1)

        self.loadHelp()
        self.loadButtons()
        self.onFirstExecBegin.append(self.afterLayout)

    def afterLayout(self):
        self.setTextFont
        self.loadStaticConfig()
        self.loadUrl(config.plugins.CurlyTx.defaultPage.value)

    def loadHelp(self):
        self.helpList.append((
                self["actions"], "WizardActions",
                [("up", _("Scroll page contents up"))]))
        self.helpList.append((
                self["actions"], "WizardActions",
                [("down", _("Scroll page contents down"))]))
        self.helpList.append((
                self["actions"], "InfobarEPGActions",
                [("showEventInfo", _("Show HTTP response headers"))]))
        self.helpList.append((
                self["actions"], "ColorActions",
                [("red", _("Show program settings"))]))
        self.helpList.append((
                self["actions"], "ColorActions",
                [("green", _("Reload current page URL"))]))
        self.helpList.append((
                self["actions"], "ColorActions",
                [("yellow", _("Switch to next configured page URL"))]))
        self.helpList.append((
                self["actions"], "ColorActions",
                [("blue", _("Switch to previous configured page URL"))]))
        self.helpList.append((
                self["actions"], "WizardActions",
                [("ok", _("Close window"))]))
        self.helpList.append((
                self["actions"], "WizardActions",
                [("back", _("Close window"))]))

    def loadButtons(self):
        pageCount = len(config.plugins.CurlyTx.pages)
        if pageCount == 0:
            self["key_green"].setText("")
            self["key_yellow"].setText("")
            self["key_blue"].setText("")
        elif pageCount == 1:
            self["key_green"].setText(_("Reload"))
            self["key_yellow"].setText("")
            self["key_blue"].setText("")
        else:
            self["key_green"].setText(_("Reload"))
            self["key_yellow"].setText(_("Prev"))
            self["key_blue"].setText(_("Next"))

        if config.plugins.CurlyTx.enableSettings.getValue():
            self["key_red"].setText(_("Settings"))
        else:
            self["key_red"].setText("")

    def pageUp(self):
        self["text"].pageUp()

    def pageDown(self):
        self["text"].pageDown()

    def prevPage(self):
        if self.currentPage == None:
            return

        pageId = self.currentPage - 1
        if pageId < 0:
            pageId = len(config.plugins.CurlyTx.pages) - 1
        self.loadUrl(pageId)

    def nextPage(self):
        if self.currentPage == None:
            return

        pageId = self.currentPage + 1
        if pageId > len(config.plugins.CurlyTx.pages) - 1:
            pageId = 0
        self.loadUrl(pageId)

    def reload(self):
        if self.currentPage == None:
            return

        self.loadUrl(self.currentPage)

    def loadUrl(self, pageId):
        self.httpGetterFactory = None
        if pageId == None:
            self.loadNoPage()
            return

        cfg = config.plugins.CurlyTx
        pageCount = len(cfg.pages)
        pageId = int(pageId)
        if pageId > (pageCount - 1):
            if len(cfg.pages) == 0:
                self.loadNoPage()
            else:
                self["text"].setText(_("Invalid page") + " " + pageId);
            return

        url   = cfg.pages[pageId].uri.value
        title = cfg.pages[pageId].title.value

        if pageCount > 1:
            title = "{0} [{1}/{2}]".format(title, pageId + 1, pageCount)

        self.currentPage = pageId
        self.currentUrl = url
        self.currentFontSize = cfg.pages[pageId].fontSize.value

        self.setTitle(title)
        self.setTextFont()
        self["text"].setText(_("Loading ...") + "\n" + url);

        if (url.startswith('file://')):
            self.loadLocalFile(url)
        else:
            self.getPageWebClient(url).addCallback(self.urlLoaded).addErrback(self.urlFailed, url)

    def loadLocalFile(self, url):
        file = url[7:]
        if not os.path.exists(file):
            self.showFail('File does not exist', file)
            return

        with open(file, 'r') as f:
            self.urlLoaded(f.read())

    def setTextFont(self):
        if self["text"].long_text is not None:
            self["text"].long_text.setFont(gFont("Console", self.currentFontSize))

    def urlLoaded(self, html):
        self["text"].setText(html)

    def urlFailed(self, error, url):
        self.showFail(error.getErrorMessage(), url)

    def showFail(self, message, url):
        self["text"].setText(
            _("Error fetching URL:") + "\n " + message
            + "\n\nURL: " + url
            )

    def loadNoPage(self):
        self["text"].setText(_("Go and add a page in the settings"));

    def showHeader(self):
        if not self.httpGetterFactory:
            return

        if self.showingHeaders:
            self["text"].setText(self.pageContent)
            self.pageContent    = None
            self.showingHeaders = False
        elif self.httpGetterFactory.response_headers:
            headers = _("HTTP response headers for") + "\n" + self.currentUrl + "\n\n"
            for (k, v) in self.httpGetterFactory.response_headers.items():
                headers += k + ": " + ("\n" + k + ": ").join(v) + "\n"
            self.pageContent = self["text"].getText()
            self["text"].setText(headers)
            self.showingHeaders = True

    def showSettings(self):
        if not config.plugins.CurlyTx.enableSettings.getValue():
            return

        from CurlyTxSettings import CurlyTxSettings
        self.session.openWithCallback(self.onSettingsChanged, CurlyTxSettings)

    def onSettingsChanged(self):
        self.loadButtons()
        if len(config.plugins.CurlyTx.pages) == 0:
            self.currentPage = None
            self.loadUrl(self.currentPage)
        elif self.currentPage == None:
            self.currentPage = 0
            self.loadUrl(self.currentPage)


    def getPageWebClient(self, url, contextFactory=None, *args, **kwargs):
        """
        Download a web page as a string.

        COPY OF twisted.web.client.getPage to store the factory

        Download a page. Return a deferred, which will callback with a
        page (as a string) or errback with a description of the error.

        See L{HTTPClientFactory} to see what extra arguments can be passed.
        """
        self.httpGetterFactory = _makeGetterFactory(
            url,
            HTTPClientFactory,
            contextFactory=contextFactory,
            *args, **kwargs)
        return self.httpGetterFactory.deferred

    def loadStaticConfig(self):
        """
        Always try to load the static config file from
        /etc/enigma2/curlytx-pagefeed.xml
        """
        staticFeedPath = Directories.resolveFilename(Directories.SCOPE_CONFIG, self.staticPageFeedFile)
        if not os.path.exists(staticFeedPath):
            return

        from AtomFeed import AtomFeed
        AtomFeed(
            'file://' + staticFeedPath,
            self.saveStaticConfig, self.loadStaticConfigFail
            )

    def loadStaticConfigFail(self, errorMessage):
        """ Loading the page url feed failed somehow """
        from Screens.MessageBox import MessageBox
        self.session.open(
            MessageBox,
            _("Error loading page feed:") + "\n\n" + str(errorMessage),
            type = MessageBox.TYPE_ERROR
            )

    def saveStaticConfig(self, pages, settings):
        feedPagesToConfig(pages)
        feedSettingsToConfig(settings)
        savePageConfig()
        self.loadButtons()
